RO-13: Planlegging

jorgkv opened the pull request:
Husk å rebase denne jevnlig så vi ikke får så mye konflikter til slutt.
Hvis du vil teste import av GPX-spor, kan du bruke f.eks. https://gpx.studio/ til å lage GPS-filer. https://geojson.io/ kan brukes til å lage GeoJSON-filer.

TODO

- [x] Vi støtter bare lat/lon i appen, trenger vi sjekke om importerte geojson-filer har koordinater som ligner på andre ting? (gpx støtter bare lat lon)
- [x] Beregne lengde på tur under import / lagring?
- [x] Fjern hardkodede turer i kartet (de som ligger i Asker)
- [x] Fjerne logging vi ikke trenger lengre (men som kan være nyttig å ha der under testing)
- [x] Kjøre copilot-review
- [ ] Sjekke hvor mye plass vi bruker i LocalStorage med 4-5 kladder + en god del turer importert. Hvis vi nærmer oss grensa på 5 mb? så burde vi kanskje hatt en slags håndtering av det, at vi melder i fra om at maks antall turer er importert. Tipper at det skal mye til for å nå den grensa, men siden dette med turer potensielt kan overskrive kladder dersom man når plassgrensa ? så burde vi undersøke det.
- [x] Gjøre det lettere å velge et spor i kartet, for å se navn og beskrivelse. Det virker som "hover-området" for sporet er for smalt (fikses i #903)
- [x] Detaljsida for tur er ikke veldig tilpassa for web, kartet kunne tatt opp (nesten) hele skjermen, og tur-metadata bare tatt opp en liten bit på siden av kartet.
- [x] Oppdatere til siste versjon av designsystemet
- [x] Vise feilmelding hvis vi ikke får til å importere en fil
- [x] Lista er vanskelig å scrolle i Android. Det funker ikke som de andre sidene. Kan ikke bare "dra" sida opp eller ned, jeg må treffe scrollbar som er veldig smal
- [ ] Tag "Vises i kart" i lista blir kuttet og helt borte når navn på spor er langt
- [x] det er noe rart med punkt-ikonene:
<img width="62" height="67" alt="image" src="https://github.com/user-attachments/assets/6e647537-0117-4a2e-b75d-bdc529861cf0" />

gruble committed:
RO-3016 og RO-3015 Lagt til NVE Designsystem og ny fane for planer (#856)

Co-authored-by: Jørgen Loe Kvalberg <jolokv@nve.no>

jorgkv committed:
RO-3017: Vise GeoJSON i kartet

Co-authored-by: Øystein Myhre <71138449+gruble@users.noreply.github.com>

jorgkv committed:
Legg til filimport med filedrop

jorgkv committed:
Konverter gpx til geojson og lagre i db

jorgkv committed:
Rydd opp import-kode

jorgkv committed:
Fiks visning på mobil

jorgkv committed:
RO-3033: listevisning av turer (#869)

Co-authored-by: marcin <martin.joodd@gmail.com>

gruble committed:
RO-3032: Vise importerte turer i kartet (#865)

gruble committed:
RO-3031: Import av GeoJSON-filer (#867)

jorgkv committed:
Legg til vasking av koordinater og egenskaper (#864)

jorgkv committed:
RO-3034: Detaljside (#870)

Co-authored-by: marcin <martin.joodd@gmail.com>

amish1188 committed:
RO-3072: beregne og vise tur lengde (#894)

Co-authored-by: Øystein Myhre <71138449+gruble@users.noreply.github.com>
Co-authored-by: Jørgen Loe Kvalberg <jolokv@nve.no>

amish1188 committed:
RO-3046: filter på tur planer visning -alle eller kun synlige i kart (#897)

gruble committed:
RO-3069: Fikset småting på planer (#898)

jorgkv commented on src/app/core/services/geojson/geojson.service.ts:
```diff
@@ -0,0 +1,127 @@
+   * @returns lengde i kilometer
+   */
+  private calculateLengthKm(geojson: Feature<Geometry, GeoJsonProperties>[]): number {
+    return geojson.reduce((sum, feature) => sum + length(feature), 0);
```
Vi har en egen fil (geojson.ts i plan-mappa) der slike ting kan ligge, da blir det enklere å enhetsteste de osv. Tenker vi bør slike ting enten der eller her i alle fall.

jorgkv commented on src/app/core/services/geojson/geojson.service.ts:
```diff
@@ -73,6 +83,13 @@ export class GeoJSONService {
+      const lineFeatures = geojson.features.filter((f) => f.geometry.type === 'LineString');
+
+      if (lineFeatures.length) {
+        const lengthKm = this.calculateLengthKm(lineFeatures);
```
Vi bør kanskje ha lengde-beregningen i en egen try-catch, eller justere på feilmeldingen under, så det ikke blir forvirrende?

jorgkv commented on src/app/pages/plans/plan/plan.page.ts:
```diff
@@ -0,0 +1,64 @@
+      lengthKm: this.itemMetadata()?.lengthKm,
+      date: Date.now(),
+    };
+    await this.geoJSON.updateMetadata(itemToUpdate);
```
Denne skal vel ikke være async? Tror onSave-metoden ikke trenger å være async, vi trenger ikke vente på router-kallet.

jorgkv commented on src/app/modules/map/components/map/map.component.ts:
```diff
@@ -450,6 +475,24 @@ export class MapComponent implements OnInit, OnDestroy, AfterViewInit {
 
+    // Lytt på endringer i evt. geojson-metadata og tegn geojson-lag (på nytt)
+    this.geoJSONService.metadata$.pipe(takeUntil(this.ngDestroy$)).subscribe(async (metadataForAllTracks) => {
+      if (metadataForAllTracks && map) {
```
Jeg mener vi bør prøve å skrive om dette før vi tar inn #857 , eller teste hvordan det funker med mange importerte geojson-filer (feks 50-100), og toggle lag av og på, lagre endringer osv.
At subscribe-funksjonen er async, gjør det litt uforutsigbart hva som skjer her syns jeg. Jeg tipper at det vil kunne skje at denne kalles flere ganger samtidig dersom endringene emites ofte nok.

Det er mange måter det kunne vært skrevet om på tenker jeg, feks ved å bare emitte endringer per lag i stedet for hele metadataen hver gang. 
Det ville også blitt litt tryggere hvis vi feks bare gjør denne oppdateringen når det ikke har kommet noen endringer på 1 sek feks.

jorgkv commented on src/app/pages/plans/plan/plan.page.html:
```diff
@@ -0,0 +1,48 @@
+          [label]="'PLANS.COMMENT' | translate"
+        ></nve-textarea>
+        <div class="button-group">
+          <nve-button variant="primary" type="submit" (click)="onSave()">Lagre</nve-button>
```
Dette med egen lagre-knapp per skjema er noe nytt i appen. Selv syns jeg ikke brukeropplevelsen er så god nå, med lagreknappen. Jeg prøvde feks å toggle en tur på bare med togglen, men glemte å trykke lagre fordi det er uvant og litt uvanlig sammenligna med appen ellers.

Tenker vi bør skrive om skjemaet til å lagre oppdateringene fortløpende så vi kan fjerne lagre-knappen. Eventuelt burde vi kanskje hatt to visninger av skjemaet, for å gjøre det tydeligere at "nå redigerer du". Feks at det var en helt statisk visning uten inputs / toggle, men med en rediger-knapp som åpner redigeringsvisning.

amish1188 replied:
vi har jo submit knapper noen steder (se bildet under). Å ikke ha en dedikert submit-knapp i et skjema gir lite mening med tanke på universell utforming. Uten en submit-knapp har vi ingen måte å informere brukeren om at noe faktisk er lagret. hvordan vil du informere skjermleser brukere at data var lagret? Nå som vi skal bruke designsystemet, må vi også ta vare på uu, syns jeg, og bruke best praksis. 
Jeg merker vi ikke har submit knapper i underskjemaer (tror dette kan diskuteres), men når du faktisk vil sende et skjema klikker du på en knapp som lagrer det for deg. Akkurat som her. Det er kun et skjema i detaljside. Derfor submit knapp gir mening. 
<img width="587" height="446" alt="image" src="https://github.com/user-attachments/assets/8e39013c-b1bf-475c-9b0d-ac2a5a46958b" />

amish1188 replied:
vi har også fjernet switch fra skjemaet. switch er ment til å lagre endring med en gang du trykker på den uten noe ekstra lagre knapp. men vi trodde også at switch vi har er forvirrende og vi fjernet den. det kan også diskuteres senere

jorgkv commented on src/app/core/services/database/database.service.ts:
```diff
@@ -55,6 +55,7 @@ export class DatabaseService {
    * @returns Returns a promise with the value of the given key
    */
   async get<T>(key: string): Promise<T> {
+    // TODO: Legg til null på type
```
Dette er litt urelatert til denne PRen. Vi kunne tatt det i en egen PR eventuelt og slette denne kommentaren. Problemet er at `this.database.get` kan returnere null dersom nøkkelen ikke finnes, men det kommer ikke fram av retur-typen som er `Promise<T>`. Men siden denne servicen brukes en del steder som ikke er relatert til denne PRen, blir det rotete å ta med endringen her.

gruble replied:
Jeg laget en oppgave på det:
https://nveprojects.atlassian.net/browse/RO-3118

gruble committed:
RO-3047: Mer effektiv tegning av spor (#902)

gruble committed:
Mer musevennlig visning av navn og beskrivelse (#903)

gruble committed:
Småfiks av planer etter kommentarer på PR (#904)

gruble committed:
Fiks for valg av sporfiler på Android (#906)

jorgkv committed:
Oppdater csp-regler for at designsystem skal funke (#907)

jorgkv committed:
Flytt planer til sidemenyen

  - Endrer ikon så det fungerer med ion-icon og farger
  - Fjerner oversettelse ikke i bruk
  - Fikser headere og back-buttons
  - Flytter routing til app.routes

amish1188 committed:
RO-3036: vise kart i plan detalj siden (#899)

jorgkv committed:
Temp commit

jorgkv committed:
Forbedringer i henhold til skissene

jorgkv committed:
Fix tests

jorgkv committed:
Juster på melding

jorgkv committed:
Bruk samme stil i kart og under plan

jorgkv committed:
Oppdater infotekst

jorgkv committed:
Fiks problem med scrolling

gruble commented on src/global.scss:
```diff
@@ -55,6 +59,40 @@ a {
 }
 
+/* Source Sans Pro brukes av NVE designsystem. TODO: Er ikke helt sikker på om vi trenger italic-variantene */
+@font-face {
```
Trenger vi dette lengre nå om vi bruker siste versjon av Designsystemet? Jeg trodde fonter ble lastet automatisk fra designsystem-pakka siden denne commiten: https://github.com/NVE/Designsystem/commit/b8186dee3eb86a9809771e8de110fb78aea98cd5 
Men det bør sikkert testes at det funker, også offline.

jorgkv committed:
Oppdater designsystem og fiks review comments

Fjerner endringer i CSP. Fjerner ubrukte skrifttyper. Legger til oversettelser som manglet. Forbedrer feilhåndtering ved filopplasting.

jorgkv committed:
Oppdater tekst.

jorgkv committed:
Fiks margin i bunnen av plans page

jorgkv committed:
Fix side menu bugs
